### PR TITLE
Out of Band protocol

### PR DESCRIPTION
**Pull request by mirgee**

### CODE REVIEW

File: aries_vcx/src/handlers/out_of_band/sender/sender.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    pub fn append_service(mut self, service: ServiceResolvable) -> Self {
+        self.services.push(service);
+        self
+    }
+
+    pub fn append_handshake_protocol(mut self, protocol: HandshakeProtocol) -> VcxResult<Self> {
+        let new_protocol = match protocol {
+            HandshakeProtocol::ConnectionV1 => MessageType::build(MessageFamilies::Connections, ""),
+            HandshakeProtocol::DidExchangeV1 => { return Err(VcxError::from(VcxErrorKind::ActionNotSupported)) }
```
Comment (Patrik-Stas) [inline aries_vcx/src/handlers/out_of_band/sender/sender.rs:37]: Good reminder of protocol work ahead of us :-) Went ahead and created an issue to track this
https://github.com/hyperledger/aries-vcx/issues/346

File: aries_vcx/tests/test_integration.rs [hunk 2]
```diff
@@ -1001,38 +1014,104 @@ mod tests {
...
+            .set_goal("To exchange message")
+            .append_service(ServiceResolvable::FullService(service));
+        let oob_msg = oob_sender.to_a2a_message();
 
         consumer.activate().unwrap();
-        let consumer_msgs = consumer_to_institution.download_messages(Some(vec![MessageStatusCode::Received]), None).unwrap();
-        assert_eq!(consumer_msgs.len(), 1);
+        let oob_receiver = OutOfBand::create_from_a2a_msg(&oob_msg).unwrap();
+        let conn = oob_receiver.connection_exists(vec![&consumer_to_institution]).unwrap();
+        assert!(conn.is_some());
```
Comment (Patrik-Stas) [inline aries_vcx/tests/test_integration.rs:1114]: ideally let's have Alice create more than 1 connection and make sure that `connection_exists` picks the right one. 
We should also test it returns None, if no matching one is supplied

Comment (Patrik-Stas) [reply aries_vcx/tests/test_integration.rs:1114]: We could also test the connection will not match if the original connection was created "privately", instead of  public did invite.

File: aries_vcx/tests/test_integration.rs [hunk 3]
```diff
@@ -1001,38 +1014,104 @@ mod tests {
...
+            .append_a2a_message(request_sender.to_a2a_message()).unwrap();
+        let oob_msg = oob_sender.to_a2a_message();
 
         consumer.activate().unwrap();
-        let mut consumer_to_institution = Connection::create_with_invite("institution", public_invite, true).unwrap();
-        consumer_to_institution.connect().unwrap();
-        consumer_to_institution.update_state().unwrap();
+        let oob_receiver = OutOfBand::create_from_a2a_msg(&oob_msg).unwrap();
+        let conn = oob_receiver.connection_exists(vec![]).unwrap();
+        assert!(!conn.is_some());
```
Comment (Patrik-Stas) [inline aries_vcx/tests/test_integration.rs:1056]: `is_none()`

File: aries_vcx/src/handlers/out_of_band/receiver/receiver.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+impl OutOfBand {
+    pub fn create_from_a2a_msg(msg: &A2AMessage) -> VcxResult<Self> {
+        match msg {
+            A2AMessage::OutOfBand(oob) => Ok(oob.clone()),
+            _ => Err(VcxError::from(VcxErrorKind::InvalidMessageFormat))
+        }
+    }
+
+    pub fn connection_exists<'a>(&self, connections: Vec<&'a Connection>) -> VcxResult<Option<&'a Connection>> {
```
Comment (Patrik-Stas) [inline aries_vcx/src/handlers/out_of_band/receiver/receiver.rs:22]: I think before resolving `full_service` of the oob, we should just try to match by DID itself - if available on connection ofcourse.
With current code, if:
- i establish connection via public invite, using their did `sov:12345`
-  institutions updates their `service` on ledger for their did `sov:12345`
- I find scan their OOB message which contains `services: ["sov:12345"]`

the code would determine these connection not to match. Which is not desired behaviour I think. Also, this scenario should be covered by a test

Comment (Patrik-Stas) [reply aries_vcx/src/handlers/out_of_band/receiver/receiver.rs:22]: Also I've realized I am not quite sure how well the reference return will actually work once we try to expose these features on handle/ffi layer. You'll have to find out the handle of the returned reference, so we could return that handle via FFI. But I guess we'll find out once we start implementing that.

File: aries_vcx/tests/test_integration.rs [hunk 5]
```diff
@@ -1001,38 +1014,104 @@ mod tests {
...
+            .append_handshake_protocol(HandshakeProtocol::ConnectionV1).unwrap()
+            .append_a2a_message(request_sender.to_a2a_message()).unwrap();
+        let oob_msg = oob_sender.to_a2a_message();
 
         consumer.activate().unwrap();
-        let mut consumer_to_institution = Connection::create_with_invite("institution", public_invite, true).unwrap();
-        consumer_to_institution.connect().unwrap();
-        consumer_to_institution.update_state().unwrap();
+        let oob_receiver = OutOfBand::create_from_a2a_msg(&oob_msg).unwrap();
+        let conn = oob_receiver.connection_exists(vec![]).unwrap();
```
Comment (Patrik-Stas) [inline aries_vcx/tests/test_integration.rs:1055]: there's not much point calling this if you only pass in empty vector though

File: aries_vcx/tests/test_integration.rs [hunk 6]
```diff
@@ -1001,38 +1014,104 @@ mod tests {
...
+        conn_sender.update_state().unwrap();
+        assert_eq!(ConnectionState::Inviter(InviterState::Completed), conn_sender.get_state());
 
-        institution_to_consumer.send_generic_message("Hello Alice, Faber here").unwrap();
+        let conn = oob_receiver.connection_exists(vec![&conn_receiver]).unwrap();
+        assert!(conn.is_some());
+
+        let a2a_msg = oob_receiver.extract_a2a_message().unwrap().unwrap();
+        assert!(matches!(a2a_msg, A2AMessage::PresentationRequest(..)));
+        if let A2AMessage::PresentationRequest(request_receiver) = a2a_msg {
```
Comment (Patrik-Stas) [inline aries_vcx/tests/test_integration.rs:1087]: I suggest only using BasicMessage for this test, not a big deal but:
- no need to create schema and cred def
- it's clearer this test is not testing oob+proofs, as it might appear at quick glance

File: aries_vcx/tests/test_integration.rs [hunk 7]
```diff
@@ -1001,38 +1014,104 @@ mod tests {
...
+            .set_goal_code(GoalCode::P2PMessaging)
+            .set_goal("To exchange message")
+            .append_service(ServiceResolvable::FullService(service));
+        let oob_msg = oob_sender.to_a2a_message();
 
         consumer.activate().unwrap();
-        let consumer_msgs = consumer_to_institution.download_messages(Some(vec![MessageStatusCode::Received]), None).unwrap();
-        assert_eq!(consumer_msgs.len(), 1);
+        let oob_receiver = OutOfBand::create_from_a2a_msg(&oob_msg).unwrap();
+        let conn = oob_receiver.connection_exists(vec![&consumer_to_institution]).unwrap();
```
Comment (Patrik-Stas) [inline aries_vcx/tests/test_integration.rs:1113]: I'd like to see BasicMessage exchanged to verify the connection is truly reused and message arrives where expected

File: aries_vcx/tests/test_integration.rs [hunk 8]
```diff
@@ -1001,38 +1014,104 @@ mod tests {
...
+        assert!(matches!(a2a_msg, A2AMessage::PresentationRequest(..)));
+        if let A2AMessage::PresentationRequest(request_receiver) = a2a_msg {
+            assert_eq!(request_receiver.thread.unwrap().pthid.unwrap(), oob_receiver.id.0);
+            assert_eq!(request_receiver.request_presentations_attach, request_sender.request_presentations_attach);
+        }
+    }
+
+    #[test]
+    #[cfg(feature = "agency_pool_tests")]
+    fn test_connection_reuse() {
```
Comment (Patrik-Stas) [inline aries_vcx/tests/test_integration.rs:1095]: I suggest rename: `test_oob_connection_reuse`

File: aries_vcx/tests/test_integration.rs [hunk 9]
```diff
@@ -1001,38 +1014,104 @@ mod tests {
...
+        institution_to_consumer.send_generic_message("Hello Alice, Faber here").unwrap();
+
+        consumer.activate().unwrap();
+        let consumer_msgs = consumer_to_institution.download_messages(Some(vec![MessageStatusCode::Received]), None).unwrap();
+        assert_eq!(consumer_msgs.len(), 1);
+    }
+
+    #[test]
+    #[cfg(feature = "agency_pool_tests")]
+    fn test_establish_connection_via_oob_message_request_included() {
```
Comment (Patrik-Stas) [inline aries_vcx/tests/test_integration.rs:1028]: I suggest rename: `test_oob_connection_bootstrap`

File: aries_vcx/tests/test_integration.rs [hunk 10]
```diff
@@ -1001,38 +1014,104 @@ mod tests {
...
         thread::sleep(Duration::from_millis(500));
-        let mut conn_requests = institution.agent.download_connection_requests(None).unwrap();
-        assert_eq!(conn_requests.len(), 1);
-        let mut institution_to_consumer = Connection::create_with_connection_request(conn_requests.pop().unwrap(), &institution.agent).unwrap();
-        assert_eq!(ConnectionState::Inviter(InviterState::Requested), institution_to_consumer.get_state());
-        institution_to_consumer.update_state().unwrap();
-        assert_eq!(ConnectionState::Inviter(InviterState::Responded), institution_to_consumer.get_state());
+        let mut requests = institution.agent.download_connection_requests(None).unwrap();
+        assert_eq!(requests.len(), 1);
+        let mut conn_sender = Connection::create_with_connection_request(requests.pop().unwrap(), &institution.agent).unwrap();
```
Comment (Patrik-Stas) [inline aries_vcx/tests/test_integration.rs:1067]: Seems to be decent fragment of code duplicated in following lines and in `create_connected_connections_via_public_invite` - I think if `create_connected_connections_via_public_invite` is perhaps split in 2 functions / generalized, we could get rid of the duplicity. 

Following lines seems to be duplicate:
```
 let mut requests = institution.agent.download_connection_requests(None).unwrap();
        assert_eq!(requests.len(), 1);
        let mut conn_sender = Connection::create_with_connection_request(requests.pop().unwrap(), &institution.agent).unwrap();
        assert_eq!(ConnectionState::Inviter(InviterState::Requested), conn_sender.get_state());
        conn_sender.update_state().unwrap();
        assert_eq!(ConnectionState::Inviter(InviterState::Responded), conn_sender.get_state());

        consumer.activate().unwrap();
        thread::sleep(Duration::from_millis(500));
        conn_receiver.update_state().unwrap();
        assert_eq!(ConnectionState::Invitee(InviteeState::Completed), conn_receiver.get_state());

        institution.activate().unwrap();
        thread::sleep(Duration::from_millis(500));
        conn_sender.update_state().unwrap();
        assert_eq!(ConnectionState::Inviter(InviterState::Completed), conn_sender.get_state());
        ```
        
Not a big deal, consider some extraction - if it turns out problematic for some reason, or not very useful, nevermind.

**Review by Patrik-Stas — CHANGES_REQUESTED** [review]
left a few comments

File: aries_vcx/tests/test_integration.rs (hunk 5 above)
Comment (mirgee) [reply aries_vcx/tests/test_integration.rs:1055]: Why? It is a perfectly valid input

File: aries_vcx/src/handlers/out_of_band/receiver/receiver.rs (hunk 4 above)
Comment (mirgee) [reply aries_vcx/src/handlers/out_of_band/receiver/receiver.rs:22]: Sure this should be done, we discussed this scenario yesterday. There are a few potential problems:
* the `id` field of a service is [not necessarily a did](https://w3c.github.io/did-core/#services)
* the original connection should be updated before reusing it
* we do not distinguish between connections created via public and pairwise did (could be fixed by distinguishing between pw and public dids using a prefix - as we should)

I agree with potential problems incoming regarding the lifetimes. We require the returned connection to live as long as every connection in the vector, which is probably not very practical.

File: aries_vcx/tests/test_integration.rs (hunk 7 above)
Comment (mirgee) [reply aries_vcx/tests/test_integration.rs:1113]: Sure thing, that's the plan

File: aries_vcx/tests/test_integration.rs (hunk 6 above)
Comment (mirgee) [reply aries_vcx/tests/test_integration.rs:1087]: True (I did that just to generate a reasonable presentation request)

File: aries_vcx/tests/test_integration.rs (hunk 2 above)
Comment (mirgee) [reply aries_vcx/tests/test_integration.rs:1114]: That's right. Ideally, I think that testing the functionality of `connection_exists` specifically should be a job of a couple unit tests.

File: aries_vcx/tests/test_integration.rs (hunk 10 above)
Comment (mirgee) [reply aries_vcx/tests/test_integration.rs:1067]: Yeah, agreed - there quite a bit of duplication in the integration tests in general.

File: aries_vcx/src/handlers/out_of_band/receiver/receiver.rs (hunk 4 above)
Comment (Patrik-Stas) [reply aries_vcx/src/handlers/out_of_band/receiver/receiver.rs:23]: > the original connection should be updated before reusing it

Assuming you maintain association `<public_did> : <vcx_connection_object>`, and you bump into the same public_did in some OOB message - you match it, doesn't mean you need to update service of connection's ddo.

> the bootstrap did doc does not contain public did currently and doing so would probably require assuming the id field of a service

That's right, `@id` can be any URI, not necessarily a DID. Quote from the spec you referred to: 
> The value of the id property MUST be a URI conforming to [RFC3986].

So I think need to remember public DID somewhere aside from DDO. I think on our Connection object level.

Comment (Patrik-Stas) [reply aries_vcx/src/handlers/out_of_band/receiver/receiver.rs:23]: > the original connection should be updated before reusing it

Assuming you maintain association `<public_did> : <vcx_connection_object>`, and you bump into the same public_did in some OOB message - you match it, doesn't mean you need to update service of connection's ddo.

> the bootstrap did doc does not contain public did currently and doing so would probably require assuming the id field of a service

That's right, `@id` can be any URI, not necessarily a DID. Quote from the spec you referred to: 
> The value of the id property MUST be a URI conforming to [RFC3986].

So I think need to remember public DID somewhere aside from DDO. I think on our Connection object level.

**Review by Patrik-Stas — APPROVED** [review]
